src/channel/mod.rs

```diff
@@ -1,14 +1,89 @@
+pub mod channel_states;
+
+pub mod fs_verifier_channel;
+pub mod fs_prover_channel;
+
 use ark_ff::Field;
+use channel_states::ChannelStates;
+use std::convert::{AsMut, AsRef};
+use crate::hashutil::TempHashContainer;
 
 #[allow(dead_code)]
-trait Channel {
+trait Channel: AsRef<ChannelStates> + AsMut<ChannelStates> {
     type Field: Field;
 
-    fn recv_felts(&mut self, n: usize) -> Result<Vec<Self::Field>, anyhow::Error>;
+    fn recv_felem(&mut self, felem: Self::Field) -> Result<Self::Field, anyhow::Error>;
 
     fn recv_bytes(&mut self, n: usize) -> Result<Vec<u8>, anyhow::Error>;
 
     fn random_number(&mut self, bound: u64) -> u64;
 
     fn random_field(&mut self) -> Self::Field;
+
+    /// Only relevant for non-interactive channels. Changes the channel seed to a "safer" seed.
+    ///
+    /// This function guarantees that randomness fetched from the channel after calling this function
+    /// and before sending data from the prover to the channel, is "safe": A malicious
+    /// prover will have to perform 2^security_bits operations for each attempt to randomize the fetched
+    /// randomness.
+    ///
+    /// Increases the amount of work a malicious prover needs to perform, in order to fake a proof.
+    #[allow(unused_variables)]
+    fn apply_proof_of_work(&mut self, security_bits: usize) -> Result<(), anyhow::Error> {
+        Err(anyhow::Error::msg("Not a fs-channel"))
+    }
+
+    fn is_end_of_proof(&self) -> bool;
+
+    fn is_query_phase(&self) -> bool {
+        AsRef::<ChannelStates>::as_ref(self).is_query_phase
+    }
+
+    fn begin_query_phase(&mut self) {
+        AsMut::<ChannelStates>::as_mut(self).is_query_phase = true;
+    }
+
+    fn increment_byte_count(&mut self, n: usize) {
+        AsMut::<ChannelStates>::as_mut(self).byte_count += n;
+    }
+
+    fn increment_commitment_count(&mut self) {
+        AsMut::<ChannelStates>::as_mut(self).commitment_count += 1;
+    }
+
+    fn increment_hash_count(&mut self) {
+        AsMut::<ChannelStates>::as_mut(self).hash_count += 1;
```

Comment:
Just do `self.states.increment_hash_count` etc.

`AsMut` and `AsRef` exists to expose members of a struct to the other consumers.

Reply:
`ChannelState` has no meaning on its own and only becomes meaningful when it's part of a `Channel`. 
So I tried putting the logic for changing `ChannelState` into the `Channel` trait.

It seems there's no real need to add dependencies between these two traits. 
I've removed all of them and put everything back in its original place.